File: android/src/test/kotlin/com/newtrip/live_photo_gallery/PigeonBoundaryTest.kt
package com.newtrip.live_photo_gallery

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

/**
 * Pigeon 边界（wire）契约测试。
 *
 * Pigeon 只保证「三端枚举的序号一致」，不保证 native 内部消费的字符串正确：
 * LivePhotoGalleryPlugin.kt 底部那几个 `.wire` 扩展 / `pgXxxOf()` 函数是本次迁移
 * 唯一的手工对齐点，写错一个字面量（例如 `PgMediaFilter.VIDEO_ONLY -> "imageOnly"`）
 * 编译通过、其它测试全绿，但选择器会直接筛错媒体类型。
 *
 * 因此这里逐值钉死每个映射，并对每个枚举做穷举校验：新增枚举值时，
 * `when` 分支缺失是编译期错误，而映射表长度不匹配则在此立即失败。
 *
 * 对应的 Dart 侧同类测试见 test/pigeon_boundary_test.dart。
 */
class PigeonBoundaryTest {

    // ── PgMediaFilter.wire ───────────────────────────────────────────────────

    @Test
    fun `PgMediaFilter wire - 每个值映射到既有 native 字符串`() {
        assertEquals("all", PgMediaFilter.ALL.wire)
        assertEquals("imageOnly", PgMediaFilter.IMAGE_ONLY.wire)
        assertEquals("videoOnly", PgMediaFilter.VIDEO_ONLY.wire)
        assertEquals("livePhotoOnly", PgMediaFilter.LIVE_PHOTO_ONLY.wire)
    }

    @Test
    fun `PgMediaFilter wire - 穷举且互不重复`() {
        val wires = PgMediaFilter.values().map { it.wire }
        assertEquals(listOf("all", "imageOnly", "videoOnly", "livePhotoOnly"), wires)
        assertEquals(wires.size, wires.toSet().size)
    }

    @Test
    fun `PgMediaFilter wire - 全部能被 PickerConfig 的 normalizeFilter 接受`() {
        // wire 字符串若拼错，PickerConfig.from 会静默回落成 "all"
        PgMediaFilter.values().forEach { filter ->
            val config = PickerConfig.from(mapOf("filterConfig" to filter.wire))
            assertEquals(
                "PgMediaFilter.$filter 的 wire 值未被 PickerConfig 识别",
                filter.wire,
                config.filterConfig
            )
        }
    }

    // ── PgMediaType.wire ─────────────────────────────────────────────────────

    @Test
    fun `PgMediaType wire - 每个值映射到公开 API 字符串`() {
        assertEquals("image", PgMediaType.IMAGE.wire)
        assertEquals("video", PgMediaType.VIDEO.wire)
        assertEquals("livePhoto", PgMediaType.LIVE_PHOTO.wire)
    }

    @Test
    fun `PgMediaType wire - 穷举且互不重复`() {
        val wires = PgMediaType.values().map { it.wire }
        assertEquals(listOf("image", "video", "livePhoto"), wires)
        assertEquals(wires.size, wires.toSet().size)
    }

    // ── PgAssetSource.wire ───────────────────────────────────────────────────

    @Test
    fun `PgAssetSource wire - local network`() {
        assertEquals("local", PgAssetSource.LOCAL.wire)
        assertEquals("network", PgAssetSource.NETWORK.wire)
        assertEquals(listOf("local", "network"), PgAssetSource.values().map { it.wire })
    }

    // ── PgExportFormat.wire ──────────────────────────────────────────────────

    @Test
    fun `PgExportFormat wire - image video livePhotoVideo`() {
        assertEquals("image", PgExportFormat.IMAGE.wire)
        assertEquals("video", PgExportFormat.VIDEO.wire)
        assertEquals("livePhotoVideo", PgExportFormat.LIVE_PHOTO_VIDEO.wire)
        assertEquals(
            listOf("image", "video", "livePhotoVideo"),
            PgExportFormat.values().map { it.wire }
        )
    }

    // ── pgMediaTypeOf ────────────────────────────────────────────────────────

    @Test
    fun `pgMediaTypeOf - 已知字符串精确映射`() {
        assertEquals(PgMediaType.IMAGE, pgMediaTypeOf("image"))
        assertEquals(PgMediaType.VIDEO, pgMediaTypeOf("video"))
        assertEquals(PgMediaType.LIVE_PHOTO, pgMediaTypeOf("livePhoto"))
    }

    @Test
    fun `pgMediaTypeOf - null 空串 大小写错误 未知值一律归一到 IMAGE`() {
        // 结果 JSON 里 mediaType 缺失时 optString 返回空串，此前会原样透给 Dart，
        // 现在统一成 image（见 CHANGELOG 的行为变化说明）
        assertEquals(PgMediaType.IMAGE, pgMediaTypeOf(null))
        assertEquals(PgMediaType.IMAGE, pgMediaTypeOf(""))
        assertEquals(PgMediaType.IMAGE, pgMediaTypeOf("Video"))
        assertEquals(PgMediaType.IMAGE, pgMediaTypeOf("live_photo"))
        assertEquals(PgMediaType.IMAGE, pgMediaTypeOf("gif"))
    }

    @Test
    fun `pgMediaTypeOf - 与 wire 互为逆映射`() {
        PgMediaType.values().forEach { type ->
            assertEquals(type, pgMediaTypeOf(type.wire))
        }
    }

    // ── pgDownloadErrorCodeOf ────────────────────────────────────────────────

    @Test
    fun `pgDownloadErrorCodeOf - 三个已知码精确映射`() {
        assertEquals(PgDownloadErrorCode.PERMISSION_DENIED, pgDownloadErrorCodeOf("PERMISSION_DENIED"))
        assertEquals(PgDownloadErrorCode.NETWORK_ERROR, pgDownloadErrorCodeOf("NETWORK_ERROR"))
        assertEquals(PgDownloadErrorCode.SAVE_FAILED, pgDownloadErrorCodeOf("SAVE_FAILED"))
    }

    @Test
    fun `pgDownloadErrorCodeOf - null 空串 未知码回退 UNKNOWN`() {
        assertEquals(PgDownloadErrorCode.UNKNOWN, pgDownloadErrorCodeOf(null))
        assertEquals(PgDownloadErrorCode.UNKNOWN, pgDownloadErrorCodeOf(""))
        assertEquals(PgDownloadErrorCode.UNKNOWN, pgDownloadErrorCodeOf("UNKNOWN"))
        assertEquals(PgDownloadErrorCode.UNKNOWN, pgDownloadErrorCodeOf("permission_denied"))
        assertEquals(PgDownloadErrorCode.UNKNOWN, pgDownloadErrorCodeOf("INVALID_ARGS"))
    }

    // ── PgPickerConfig.toPickerConfig ────────────────────────────────────────

    private fun pgConfig(
        isDarkMode: Boolean = false,
        maxCount: Long = 9,
        enableVideo: Boolean = true,
        enableLivePhoto: Boolean = true,
        showRadio: Boolean = true,
        maxVideoCount: Long = -1,
        videoMaxDuration: Double = 0.0,
        filterConfig: PgMediaFilter = PgMediaFilter.ALL,
        cropConfig: PgCropConfig? = null
    ) = PgPickerConfig(
        isDarkMode = isDarkMode,
        maxCount = maxCount,
        enableVideo = enableVideo,
        enableLivePhoto = enableLivePhoto,
        showRadio = showRadio,
        maxVideoCount = maxVideoCount,
        videoMaxDuration = videoMaxDuration,
        filterConfig = filterConfig,
        cropConfig = cropConfig
    )

    @Test
    fun `toPickerConfig - 默认值逐字段还原`() {
        val config = pgConfig().toPickerConfig()
        assertEquals(9, config.maxCount)
        assertEquals(true, config.enableVideo)
        assertEquals(true, config.enableLivePhoto)
        assertEquals(true, config.showRadio)
        assertEquals(false, config.isDarkMode)
        assertEquals(-1, config.maxVideoCount)
        assertEquals(0.0, config.videoMaxDuration, 0.0001)
        assertEquals("all", config.filterConfig)
    }

    @Test
    fun `toPickerConfig - 全部非默认值逐字段还原且不串位`() {
        val config = pgConfig(
            isDarkMode = true,
            maxCount = 3,
            enableVideo = false,
            enableLivePhoto = true,
            showRadio = false,
            maxVideoCount = 2,
            videoMaxDuration = 30.5,
            filterConfig = PgMediaFilter.LIVE_PHOTO_ONLY
        ).toPickerConfig()
        assertEquals(3, config.maxCount)
        assertEquals(false, config.enableVideo)
        assertEquals(true, config.enableLivePhoto)
        assertEquals(false, config.showRadio)
        assertEquals(true, config.isDarkMode)
        assertEquals(2, config.maxVideoCount)
        assertEquals(30.5, config.videoMaxDuration, 0.0001)
        assertEquals("livePhotoOnly", config.filterConfig)
    }

    @Test
    fun `toPickerConfig - 每个 filterConfig 枚举值都能穿透到 PickerConfig`() {
        PgMediaFilter.values().forEach { filter ->
            assertEquals(
                filter.wire,
                pgConfig(filterConfig = filter).toPickerConfig().filterConfig
            )
        }
    }

    @Test
    fun `toPickerConfig - maxCount 归一化 非正值抬到 1`() {
        assertEquals(1, pgConfig(maxCount = 0).toPickerConfig().maxCount)
        assertEquals(1, pgConfig(maxCount = -5).toPickerConfig().maxCount)
        assertEquals(1, pgConfig(maxCount = 1).toPickerConfig().maxCount)
        assertEquals(100, pgConfig(maxCount = 100).toPickerConfig().maxCount)
    }

    @Test
    fun `toPickerConfig - maxVideoCount 归一化 -1 表无限制 其余非正值抬到 1`() {
        assertEquals(-1, pgConfig(maxVideoCount = -1).toPickerConfig().maxVideoCount)
        assertEquals(1, pgConfig(maxVideoCount = 0).toPickerConfig().maxVideoCount)
        assertEquals(1, pgConfig(maxVideoCount = -3).toPickerConfig().maxVideoCount)
        assertEquals(5, pgConfig(maxVideoCount = 5).toPickerConfig().maxVideoCount)
    }

    @Test
    fun `toPickerConfig - videoMaxDuration 归一化 负值归零`() {
        assertEquals(0.0, pgConfig(videoMaxDuration = -1.0).toPickerConfig().videoMaxDuration, 0.0001)
        assertEquals(0.0, pgConfig(videoMaxDuration = 0.0).toPickerConfig().videoMaxDuration, 0.0001)
        assertEquals(15.0, pgConfig(videoMaxDuration = 15.0).toPickerConfig().videoMaxDuration, 0.0001)
    }

    @Test
    fun `toPickerConfig - Long 到 Int 的窄化不丢值`() {
        // pigeon 的整数一律是 Long，PickerConfig 用 Int；转换必须走 toInt()，
        // 否则 args map 里是 Long，`as? Int` 会失配并静默回落成默认值
        val config = pgConfig(maxCount = 7, maxVideoCount = 4).toPickerConfig()
        assertEquals(7, config.maxCount)
        assertEquals(4, config.maxVideoCount)
    }

    @Test
    fun `toPickerConfig - cropConfig 不参与 PickerConfig 且 autoPlayVideo 保持默认`() {
        // cropConfig 由预览页单独消费，不进 PickerConfig；autoPlayVideo 无对应契约字段
        val config = pgConfig(
            cropConfig = PgCropConfig(aspectRatioX = 16.0, aspectRatioY = 9.0)
        ).toPickerConfig()
        assertEquals(false, config.autoPlayVideo)
    }

    @Test
    fun `toPickerConfig - effectiveEnable 派生值随 filterConfig 生效`() {
        // wire 字符串直接驱动这两个派生属性，映射写错会在这里暴露
        val imageOnly = pgConfig(filterConfig = PgMediaFilter.IMAGE_ONLY).toPickerConfig()
        assertEquals(false, imageOnly.effectiveEnableVideo)
        assertEquals(false, imageOnly.effectiveEnableLivePhoto)

        val videoOnly = pgConfig(filterConfig = PgMediaFilter.VIDEO_ONLY).toPickerConfig()
        assertEquals(true, videoOnly.effectiveEnableVideo)
        assertEquals(false, videoOnly.effectiveEnableLivePhoto)

        val livePhotoOnly = pgConfig(filterConfig = PgMediaFilter.LIVE_PHOTO_ONLY).toPickerConfig()
        assertEquals(false, livePhotoOnly.effectiveEnableVideo)
        assertEquals(true, livePhotoOnly.effectiveEnableLivePhoto)

        val all = pgConfig(filterConfig = PgMediaFilter.ALL).toPickerConfig()
        assertEquals(true, all.effectiveEnableVideo)
        assertEquals(true, all.effectiveEnableLivePhoto)
    }

    @Test
    fun `toPickerConfig - JSON 往返后仍等价（Activity 侧通过 Intent extra 收到的形状）`() {
        val config = pgConfig(
            isDarkMode = true,
            maxCount = 6,
            enableVideo = false,
            showRadio = false,
            maxVideoCount = 2,
            videoMaxDuration = 20.0,
            filterConfig = PgMediaFilter.VIDEO_ONLY
        ).toPickerConfig()
        val restored = PickerConfig.fromJson(PickerConfig.toJson(config))
        assertEquals(config, restored)
    }

    // ── 事件侧：PgDownloadResultEvent 的失败码来源 ────────────────────────────

    @Test
    fun `pgDownloadErrorCodeOf - 覆盖 PgDownloadErrorCode 的每个取值`() {
        // UNKNOWN 没有对应的 native 字符串（是兜底），其余三个必须能被反解出来
        val reachable = listOf("PERMISSION_DENIED", "NETWORK_ERROR", "SAVE_FAILED")
            .map { pgDownloadErrorCodeOf(it) }
            .toSet()
        assertEquals(
            PgDownloadErrorCode.values().toSet() - PgDownloadErrorCode.UNKNOWN,
            reachable
        )
        assertNull(PgDownloadErrorCode.values().find { it !in reachable && it != PgDownloadErrorCode.UNKNOWN })
    }
}

File: test/pigeon_boundary_test.dart
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:live_photo_gallery/live_photo_gallery.dart';
import 'package:live_photo_gallery/src/messages.g.dart';

/// Pigeon 边界（wire）契约测试。
///
/// Pigeon 迁移后，跨端真正流动的不再是 `toMap()` / `fromMap()`，而是
/// lib/live_photo_gallery.dart 里手写的门面 ↔ 生成类型转换器：
///   MediaFilter._pg / PickerConfig._pg / AssetInput._pg /
///   MediaItem._fromPg / DownloadErrorCode._fromPg / _parseExportFormat
/// 这些转换器是本次迁移唯一的手工对齐点，一处笔误（例如把 videoOnly 映射成
/// PgMediaFilter.imageOnly）编译期无感、运行期直接选错媒体类型。
///
/// 因为它们是私有成员，本文件不去改动公开 API 来「开后门」，而是在
/// **真实的 pigeon channel 上挂 mock native**：调用公开方法 → 拦截并用生成的
/// codec 解码出 Pg* 对象 → 断言字段。这样同时覆盖了转换器 + 编解码路径，
/// 也保证生成类型不外泄到公开 API。
///
/// 对应的 native 侧同类测试见：
///   android/src/test/kotlin/com/newtrip/live_photo_gallery/PigeonBoundaryTest.kt
/// 公开门面（toMap/fromMap 等仍对调用方暴露的形状）见 test/contract_test.dart。
void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  const String hostPrefix =
      'dev.flutter.pigeon.live_photo_gallery.LivePhotoGalleryHostApi.';
  const String flutterPrefix =
      'dev.flutter.pigeon.live_photo_gallery.LivePhotoGalleryFlutterApi.';
  const MessageCodec<Object?> codec =
      LivePhotoGalleryHostApi.pigeonChannelCodec;

  final TestDefaultBinaryMessenger messenger =
      TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger;

  /// 最近一次被 mock native 接住的调用参数（已由 pigeon codec 解码）
  List<Object?> captured = const <Object?>[];

  /// 在真实 channel 上挂一个 mock native：记录入参并回复 [reply]。
  void mockHost(String method, {Object? reply}) {
    final channel =
        BasicMessageChannel<Object?>('$hostPrefix$method', codec);
    messenger.setMockDecodedMessageHandler<Object?>(channel,
        (Object? message) async {
      captured = (message as List<Object?>?) ?? const <Object?>[];
      return <Object?>[reply];
    });
  }

  void clearHost(String method) {
    final channel =
        BasicMessageChannel<Object?>('$hostPrefix$method', codec);
    messenger.setMockDecodedMessageHandler<Object?>(channel, null);
  }

  /// 模拟 native → Dart 的事件推送（走 LivePhotoGalleryFlutterApi 的真实 channel）
  Future<void> pushFlutterEvent(String method, List<Object?> args) {
    final completer = Completer<void>();
    ServicesBinding.instance.channelBuffers.push(
      '$flutterPrefix$method',
      codec.encodeMessage(args),
      (ByteData? _) => completer.complete(),
    );
    return completer.future;
  }

  /// 触发一次 pickAssets，返回 mock native 收到的 PgPickerConfig
  Future<PgPickerConfig> pgConfigOf(PickerConfig config) async {
    await LivePhotoGallery.pickAssets(config: config);
    return captured.single! as PgPickerConfig;
  }

  /// 触发一次 previewAssets，返回 mock native 收到的 PgPreviewRequest
  Future<PgPreviewRequest> pgRequestOf(List<AssetInput> assets) async {
    await LivePhotoGallery.previewAssets(assets: assets);
    return captured.single! as PgPreviewRequest;
  }

  setUp(() {
    captured = const <Object?>[];
    mockHost('pickAssets');
    mockHost('previewAssets');
    mockHost('exportAsset');
  });

  tearDown(() {
    clearHost('pickAssets');
    clearHost('previewAssets');
    clearHost('exportAsset');
  });

  // ──────────────────────────────────────────────────────────────────────────
  // PickerConfig._pg / MediaFilter._pg
  // ──────────────────────────────────────────────────────────────────────────

  group('PickerConfig._pg（公开配置 → PgPickerConfig）', () {
    test('默认配置逐字段过桥，cropConfig 为 null', () async {
      final pg = await pgConfigOf(const PickerConfig());
      expect(pg.isDarkMode, isFalse);
      expect(pg.maxCount, 9);
      expect(pg.enableVideo, isTrue);
      expect(pg.enableLivePhoto, isTrue);
      expect(pg.showRadio, isTrue);
      expect(pg.maxVideoCount, -1);
      expect(pg.videoMaxDuration, 0.0);
      expect(pg.filterConfig, PgMediaFilter.all);
      expect(pg.cropConfig, isNull);
    });

    test('全部非默认值逐字段过桥（含 cropConfig）', () async {
      final pg = await pgConfigOf(const PickerConfig(
        isDarkMode: true,
        maxCount: 3,
        enableVideo: false,
        enableLivePhoto: false,
        showRadio: false,
        maxVideoCount: 2,
        videoMaxDuration: 30.5,
        filterConfig: MediaFilter.videoOnly,
        cropConfig: CropConfig(aspectRatioX: 16, aspectRatioY: 9),
      ));
      expect(pg.isDarkMode, isTrue);
      expect(pg.maxCount, 3);
      expect(pg.enableVideo, isFalse);
      expect(pg.enableLivePhoto, isFalse);
      expect(pg.showRadio, isFalse);
      expect(pg.maxVideoCount, 2);
      expect(pg.videoMaxDuration, 30.5);
      expect(pg.filterConfig, PgMediaFilter.videoOnly);
      expect(pg.cropConfig?.aspectRatioX, 16.0);
      expect(pg.cropConfig?.aspectRatioY, 9.0);
    });
  });

  group('MediaFilter._pg（每个枚举值必须一一对应，不得串位）', () {
    test('all → PgMediaFilter.all', () async {
      final pg = await pgConfigOf(const PickerConfig(filterConfig: MediaFilter.all));
      expect(pg.filterConfig, PgMediaFilter.all);
    });

    test('imageOnly → PgMediaFilter.imageOnly', () async {
      final pg = await pgConfigOf(
          const PickerConfig(filterConfig: MediaFilter.imageOnly));
      expect(pg.filterConfig, PgMediaFilter.imageOnly);
    });

    test('videoOnly → PgMediaFilter.videoOnly', () async {
      final pg = await pgConfigOf(
          const PickerConfig(filterConfig: MediaFilter.videoOnly));
      expect(pg.filterConfig, PgMediaFilter.videoOnly);
    });

    test('livePhotoOnly → PgMediaFilter.livePhotoOnly', () async {
      final pg = await pgConfigOf(
          const PickerConfig(filterConfig: MediaFilter.livePhotoOnly));
      expect(pg.filterConfig, PgMediaFilter.livePhotoOnly);
    });

    test('公开枚举与传输枚举的取值个数一致（新增值时提醒同步）', () {
      expect(MediaFilter.values.length, PgMediaFilter.values.length);
    });
  });

  // ──────────────────────────────────────────────────────────────────────────
  // AssetInput._pg
  // ──────────────────────────────────────────────────────────────────────────

  group('AssetInput._pg（type / mediaType 归一化）', () {
    test('type 仅 "network" 判为网络源，其余（含非法值）按本地处理', () async {
      final request = await pgRequestOf(const [
        AssetInput(type: 'network', url: 'https://x/i.jpg'),
        AssetInput(type: 'local', assetId: 'a1'),
        AssetInput(type: 'NETWORK', assetId: 'a2'),
        AssetInput(type: '', assetId: 'a3'),
      ]);
      expect(
        request.assets.map((a) => a.type).toList(),
        const [
          PgAssetSource.network,
          PgAssetSource.local,
          PgAssetSource.local,
          PgAssetSource.local,
        ],
      );
    });

    test('mediaType：null 保持 null，未知字符串归一到 image', () async {
      final request = await pgRequestOf(const [
        AssetInput(type: 'local', assetId: 'a0'),
        AssetInput(type: 'local', assetId: 'a1', mediaType: 'image'),
        AssetInput(type: 'local', assetId: 'a2', mediaType: 'video'),
        AssetInput(type: 'local', assetId: 'a3', mediaType: 'livePhoto'),
        AssetInput(type: 'local', assetId: 'a4', mediaType: 'LivePhoto'),
        AssetInput(type: 'local', assetId: 'a5', mediaType: ''),
      ]);
      expect(
        request.assets.map((a) => a.mediaType).toList(),
        const [
          null,
          PgMediaType.image,
          PgMediaType.video,
          PgMediaType.livePhoto,
          PgMediaType.image,
          PgMediaType.image,
        ],
      );
    });

    test('其余字段原样过桥', () async {
      final request = await pgRequestOf(const [
        AssetInput(
          type: 'network',
          assetId: 'asset-1',
          url: 'https://x/i.jpg',
          mediaType: 'livePhoto',
          videoUrl: 'https://x/v.mov',
          duration: 2.5,
        ),
      ]);
      final asset = request.assets.single;
      expect(asset.assetId, 'asset-1');
      expect(asset.url, 'https://x/i.jpg');
      expect(asset.videoUrl, 'https://x/v.mov');
      expect(asset.duration, 2.5);
    });
  });

  group('previewAssets 请求体其余字段', () {
    test('initialIndex / sourceFrame / selectedAssetIds / 下载参数过桥', () async {
      await LivePhotoGallery.previewAssets(
        assets: const [AssetInput(type: 'local', assetId: 'a1')],
        initialIndex: 2,
        sourceFrame: const Rect.fromLTWH(1, 2, 3, 4),
        selectedAssetIds: const ['a1', 'a2'],
        showDownloadButton: true,
        saveAlbumName: '相册名',
      );
      final request = captured.single! as PgPreviewRequest;
      expect(request.initialIndex, 2);
      expect(request.sourceFrame.x, 1.0);
      expect(request.sourceFrame.y, 2.0);
      expect(request.sourceFrame.width, 3.0);
      expect(request.sourceFrame.height, 4.0);
      expect(request.selectedAssetIds, ['a1', 'a2']);
      expect(request.showDownloadButton, isTrue);
      expect(request.saveAlbumName, '相册名');
    });
  });

  // ──────────────────────────────────────────────────────────────────────────
  // MediaItem._fromPg
  // ──────────────────────────────────────────────────────────────────────────

  group('MediaItem._fromPg（PgMediaType → 公开字符串）', () {
    PgPickResult resultWith(List<PgMediaType> types) => PgPickResult(
          items: types
              .map((t) => PgMediaItem(
                    assetId: 'id-${t.name}',
                    mediaType: t,
                    thumbnailPath: '/tmp/${t.name}.jpg',
                    duration: t == PgMediaType.video ? 3.0 : null,
                    width: 100,
                    height: 200,
                  ))
              .toList(),
          isOriginalPhoto: true,
        );

    test('三种类型分别映射为 image / video / livePhoto', () async {
      mockHost('pickAssets',
          reply: resultWith(PgMediaType.values));
      final result = await LivePhotoGallery.pickAssets();
      expect(
        result!.items.map((i) => i.mediaType).toList(),
        const ['image', 'video', 'livePhoto'],
      );
      expect(result.isOriginalPhoto, isTrue);
    });

    test('其余字段原样还原', () async {
      mockHost('pickAssets', reply: resultWith(const [PgMediaType.video]));
      final item = (await LivePhotoGallery.pickAssets())!.items.single;
      expect(item.assetId, 'id-video');
      expect(item.thumbnailPath, '/tmp/video.jpg');
      expect(item.duration, 3.0);
      expect(item.width, 100);
      expect(item.height, 200);
    });

    test('native 返回 null（用户取消）时结果为 null', () async {
      mockHost('pickAssets');
      expect(await LivePhotoGallery.pickAssets(), isNull);
    });

    test('公开 mediaType 取值个数与传输枚举一致（新增值时提醒同步）', () {
      expect(PgMediaType.values.length, 3);
    });
  });

  // ──────────────────────────────────────────────────────────────────────────
  // DownloadErrorCode._fromPg（native → Dart 事件）
  // ──────────────────────────────────────────────────────────────────────────

  group('DownloadErrorCode._fromPg（PgDownloadErrorCode → 公开枚举）', () {
    setUp(() async {
      // 事件通道由 pickAssets / previewAssets 惰性注册，先触发一次
      await LivePhotoGallery.pickAssets();
    });

    Future<DownloadResult> emit(PgDownloadResultEvent event) async {
      final future = LivePhotoGallery.onDownloadResult.first;
      await pushFlutterEvent('onDownloadResult', <Object?>[event]);
      return future;
    }

    test('四个错误码逐一映射，errorCode 缺失回退 unknown', () async {
      const pairs = <(PgDownloadErrorCode?, DownloadErrorCode)>[
        (PgDownloadErrorCode.permissionDenied, DownloadErrorCode.permissionDenied),
        (PgDownloadErrorCode.networkError, DownloadErrorCode.networkError),
        (PgDownloadErrorCode.saveFailed, DownloadErrorCode.saveFailed),
        (PgDownloadErrorCode.unknown, DownloadErrorCode.unknown),
        (null, DownloadErrorCode.unknown),
      ];
      for (final (pg, expected) in pairs) {
        final result = await emit(PgDownloadResultEvent(
          url: 'https://x/i.jpg',
          success: false,
          errorCode: pg,
          errorMessage: '失败',
        ));
        expect(result, isA<DownloadFailure>());
        expect((result as DownloadFailure).errorCode, expected,
            reason: '$pg 应映射为 $expected');
        expect(result.url, 'https://x/i.jpg');
        expect(result.errorMessage, '失败');
      }
    });

    test('errorMessage 缺失时使用默认文案', () async {
      final result = await emit(PgDownloadResultEvent(
        url: 'https://x/i.jpg',
        success: false,
        errorCode: PgDownloadErrorCode.saveFailed,
      ));
      expect((result as DownloadFailure).errorMessage, '保存失败');
    });

    test('success 事件产出 DownloadSuccess 并带 assetId', () async {
      final result = await emit(PgDownloadResultEvent(
        url: 'https://x/i.jpg',
        success: true,
        assetId: 'content://media/1',
      ));
      expect(result, isA<DownloadSuccess>());
      expect((result as DownloadSuccess).assetId, 'content://media/1');
      expect(result.url, 'https://x/i.jpg');
    });

    test('onDownloadProgress / onMaxCountReached 事件过桥', () async {
      final progress = LivePhotoGallery.onDownloadProgress.first;
      await pushFlutterEvent(
          'onDownloadProgress', <Object?>['https://x/i.jpg', 0.42]);
      expect((await progress).url, 'https://x/i.jpg');
      expect((await progress).progress, 0.42);

      final maxCount = LivePhotoGallery.onMaxCountReached.first;
      await pushFlutterEvent('onMaxCountReached', <Object?>[9]);
      expect(await maxCount, 9);
    });
  });

  // ──────────────────────────────────────────────────────────────────────────
  // exportAsset 的 format 契约
  // ──────────────────────────────────────────────────────────────────────────

  group('exportAsset format → PgExportFormat', () {
    test('三个合法值分别映射', () async {
      for (final (raw, expected) in const <(String, PgExportFormat)>[
        ('image', PgExportFormat.image),
        ('video', PgExportFormat.video),
        ('livePhotoVideo', PgExportFormat.livePhotoVideo),
      ]) {
        await LivePhotoGallery.exportAsset(assetId: 'a1', format: raw);
        expect(captured, ['a1', expected]);
      }
    });

    test('非法 format 抛 INVALID_ARGS，且不会发出任何 native 调用', () async {
      for (final bad in const ['', 'Image', 'livephotovideo', 'gif']) {
        captured = const <Object?>[];
        await expectLater(
          () => LivePhotoGallery.exportAsset(assetId: 'a1', format: bad),
          throwsA(isA<LivePhotoException>()
              .having((e) => e.code, 'code', 'INVALID_ARGS')
              .having((e) => e.message, 'message', '不支持的媒体类型')),
        );
        expect(captured, isEmpty, reason: '"$bad" 不应到达 native');
      }
    });

    test('native 抛出的 PlatformException 转换为 LivePhotoException', () async {
      const channel = BasicMessageChannel<Object?>(
          '${hostPrefix}exportAsset', codec);
      messenger.setMockDecodedMessageHandler<Object?>(channel,
          (Object? message) async => <Object?>['EXPORT_FAILED', '导出失败', null]);
      await expectLater(
        () => LivePhotoGallery.exportAsset(assetId: 'a1', format: 'video'),
        throwsA(isA<LivePhotoException>()
            .having((e) => e.code, 'code', 'EXPORT_FAILED')
            .having((e) => e.message, 'message', '导出失败')),
      );
    });
  });
}
